python/cudf/cudf/core/copy_types.py

```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+from dataclasses import dataclass
+from functools import cached_property
+from typing import TYPE_CHECKING, Any, cast
+
+import cudf
+import cudf._lib as libcudf
+from cudf._lib.types import size_type_dtype
+
+if TYPE_CHECKING:
+    from cudf.core.column import NumericalColumn
+
+
+@dataclass
+class GatherMap:
+    """A witness to the validity of a given column as a gather map.
```

Comment:
This is mostly a question about terminology and semantics. Admittedly I haven't come across "witness" types like this before but perhaps they are used in other languages/contexts:

What does it mean for a `GatherMap` to be a _witness_, versus a type _composed_ of three things: an integer column of indexes, the number of rows in a target column, and a nullify parameter?

Reply:
The tuple type `(column_indices, rows_in_target, nullify)` doesn't carry the extra semantic information that `column_indices` are valid for gathering (with given `nullify` parameter) from a column with `rows_in_target` rows.

The extra piece of information you need is a proof that the indices are valid (computed by doing a reduction to check that they are in bounds). So we need to add some extra information to the tuple type to carry around this "proof" information. The `GatherMap` type is that information. It is a witness to the statement that "this tuple of indices, rows, nullify" is valid. You can think of this type as a new type or a pair of the original objects and a proof of this statement.

The terminology comes from [logic](https://en.m.wikipedia.org/wiki/Witness_(mathematics)), formally if you have some existentially qualified proposition: $\exists x \in X . \phi(x)$ then a _witness_ is some (specific) $t \in X . \phi(t)$. AKA, "you claim there is such a thing, show me it.". In this case you can think of the set $X$ as the set of all indices that are valid for gathering `n` rows (with specified `nullify` flag), and the `GatherMap` type is giving you one such $x \in X$.

The use here is I suppose a little colloquial but reasonably common in the proof assistant literature.

Reply:
https://wiki.haskell.org/Type_witness

Reply:
That is not quite this usage. Here I'm using witness in the sense of a value + proof of some restriction. e.g., in Haskell-ese:

```haskell

data NonEmptyList a = NonEmpty a [a]

head :: NonEmptyList a -> a
head (NonEmpty x _) = x

-- rather than the partial function
head' :: [a] -> Maybe a
head' (x:_) = Just x
head' _ = Nothing
```
So here `NonEmptyList` is a pair of a list, plus a proof that the list is non-empty. So the type witnesses the statement that the list is non-empty.

Reply:
Would it be equivalent "proof" if the `GatherMap` constructor did this validation? Then, the fact that you were able to construct a `GatherMap` instance is proof enough that it is valid. If that's way off, sorry! Still trying to wrap my head around this

Reply:
That's what the `as_gather_map` function does

Reply:
Performing validation in the constructor would be opposed to “parse don’t validate”. If I am understanding correctly, you should be able to construct a GatherMap without triggering validation if you know (by some prior knowledge) that the inputs are valid. That’s where the validating factory function comes in. Is that right?

Reply:
I guess it's a bit of a philosophical distinction in python (In Haskell data type constructors don't _do_ anything, so you need to use a function to do validation plus production of a datatype, in Python your init method can do arbitrary validation).

There are two things here: what do you know when you have a `GatherMap` object and how are you allowed to construct one? What you know (or at least, what the API contract promises) is that if you hold in your hand a `GatherMap(column, nrows, nullify)` object, then it is suitable for gathering a table with `nrows` rows. The idea being that by providing this object to an API that asks for a suitable gather map, you promise "yes, it is fine" (and, moreover, can distinguish that explicit promise from the implicit one that the low-level gather APIs require that you've provided valid input).

As to construction. There's nothing stopping you from just calling the constructor (I can't hide it from public view) and hence (in effect) asserting by appeal to authority that the `GatherMap` you build is valid. So my idea was to make the construction explicit in the sense that the person constructing the gather map should think about that step. So the module offers an `as_gather_map` function to do this. The type is not very different from the gather map constructor:

```python
def as_gather_map(column, nrows, nullify, check_bounds) -> GatherMap:
    ...

GatherMap(column, nrows, nullify) -> GatherMap
```

The small (possible) advantage of the separate function is that I think it makes it slightly clearer that the bounds checking (if any) is performed in construction (if I have a `GatherMap` with a `check_bounds=True` slot, does that mean that I check the bounds every time it is used in a new context, or is it as before?).

TBH it might mostly be a personal style preference here: I tend to prefer cheap-to-construct dataclass-style objects and functional interfaces to expensive object construction.

Reply:
After discussion we've gone with:

1. `GatherMap.__init__(Any, nrows: int, nullify: bool)` _constructor_ takes anything, converts to column and does validation (no opt out)
2. Class method `GatherMap.from_column_unchecked(Column, nrows: int, nullify: bool)` believes you and doesn't check anything.